### ISSUE TITLE
Questions about data preprocessing

### ISSUE THREAD

**Issue by guligualagulu** [kwanyun/StyleSketch#3]
Hello, I used the code you provided and found that only 15 sketches could be generated, because there should be only 15.pt files under dataset_release/latent. Can you tell me how to create these.pt files?

**Comment by kwanyun** [comment]
Hi, to obtain the latent code, any styleGAN2 inversion method can be used.
For our experiments, we used e4e followed by additional optimization.
[e4e](https://github.com/omertov/encoder4editing)
or [psp](https://github.com/eladrich/pixel2style2pixel), II2s, Image2StyleGAN, reStyle... etc will also work

**Comment by guligualagulu** [comment]
Glad to receive your prompt reply. I am trying it on e4e and hope it goes well.

**Comment by guligualagulu** [comment]
> 嗨，要获取潜在代码，可以使用任何styleGAN2反转方法。在我们的实验中，我们使用了 e4e，然后进行了额外的优化。 [e4e](https://github.com/omertov/encoder4editing) 或 [psp](https://github.com/eladrich/pixel2style2pixel)、II2s、Image2StyleGAN、reStyle...等等也可以

Well, it worked. But I think the result of the synthesis may be limited by how good the underlying code is. So when I use your method, there may be a little worse results.

**Comment by guligualagulu** [comment]
> Hi, to obtain the latent code, any styleGAN2 inversion method can be used. For our experiments, we used e4e followed by additional optimization. [e4e](https://github.com/omertov/encoder4editing) or [psp](https://github.com/eladrich/pixel2style2pixel), II2s, Image2StyleGAN, reStyle... etc will also work

Well, it worked. But I think the result of the synthesis may be limited by how good the latent code is. So when I use your method, there may be a little worse results.

**Comment by bikraj2** [comment]
@guligualagulu How did you achieve this? I want to create a dataset from this model to feed into the other model but been having trouble getting the latent code

**Comment by guligualagulu** [comment]
> @guligualagulu How did you achieve this? I want to create a dataset from this model to feed into the other model but been having trouble getting the latent code

I just replaced the dataset. Note that you need to specify encode only when testing.

**Comment by bikraj2** [comment]
Can you describe in detail a bit?Which dataset you replaced?

**Comment by guligualagulu** [comment]
> Can you describe in detail a bit?Which dataset you replaced?

e4e should be to obtain the encoding by reconstruction.
Follow the prompts given by the code. https://github.com/omertov/encoder4editing/tree/main#training-on-your-own-dataset.
Just choose your own dataset to train on.

**Comment by guligualagulu** [comment]
> Can you describe in detail a bit?Which dataset you replaced?

It should be noted that the training of the network requires a large amount of GPU memory, and the training takes a long time.

**Comment by bikraj2** [comment]
Thank you for the prompt response. So if I wanted latent code for say 1000 images , I would need to keep those images as my dataset?

**Comment by bikraj2** [comment]
I was using the scripts/interface.py to generate latent code but got only one latent code no matter how many images I provided

**Comment by kwanyun** [comment]
Did you check the dimensions of your latent code? For example, it might have been obtained in batches.

**Comment by guligualagulu** [comment]
> I was using the scripts/interface.py to generate latent code but got only one latent code no matter how many images I provided

You need to modify the main function in interaction.py so that each latent code is stored as code with a different name.

**Comment by bikraj2** [comment]
@kwanyun it is [4, 18, 512] does this mean it has latent space representation fo 4 images?

**Comment by bikraj2** [comment]
@guligualagulu can you guide me how I would do that

**Comment by guligualagulu** [comment]
> @guligualagulu can you guide me how I would do that

Here's my interface.py, hope it works.
import argparse

import torch
import numpy as np
import sys
import os
import dlib

sys.path.append(".")
sys.path.append("..")

from configs import data_configs, paths_config
from datasets.inference_dataset import InferenceDataset
from torch.utils.data import DataLoader
from utils.model_utils import setup_model
from utils.common import tensor2im
from utils.alignment import align_face
from PIL import Image


def main(args):
    net, opts = setup_model(args.ckpt, device)
    is_cars = 'cars_' in opts.dataset_type
    generator = net.decoder
    generator.eval()
    args, data_loader = setup_data_loader(args, opts)

    # Check if latents exist
    latents_file_path = os.path.join(args.save_dir, 'latents.pt')
    if os.path.exists(latents_file_path):
        latent_codes = torch.load(latents_file_path).to(device)
    else:
        latent_codes = get_all_latents(net, data_loader, args.n_sample, is_cars=is_cars)
        torch.save(latent_codes, latents_file_path)
        for i, codes in enumerate(latent_codes):
            filename = f"latent_{i}.pt"  # Use a different file name
            filepath = os.path.join(args.save_dir, filename)  # save path
            torch.save(codes.unsqueeze(0), filepath)  # Save the underlying code to a file
    if not args.latents_only:
        generate_inversions(args, generator, latent_codes, is_cars=is_cars)


def setup_data_loader(args, opts):
    dataset_args = data_configs.DATASETS[opts.dataset_type]
    transforms_dict = dataset_args['transforms'](opts).get_transforms()
    images_path = args.images_dir if args.images_dir is not None else dataset_args['test_source_root']
    print(f"images path: {images_path}")
    align_function = None
    if args.align:
        align_function = run_alignment
    import pdb
    pdb.set_trace()
    test_dataset = InferenceDataset(root=images_path,
                                    transform=transforms_dict['transform_test'],
                                    preprocess=align_function,
                                    opts=opts)

    data_loader = DataLoader(test_dataset,
                             batch_size=args.batch,
                             shuffle=False,
                             num_workers=2,
                             drop_last=True)

    print(f'dataset length: {len(test_dataset)}')

    if args.n_sample is None:
        args.n_sample = len(test_dataset)
    return args, data_loader


def get_latents(net, x, is_cars=False):
    codes = net.encoder(x)
    if net.opts.start_from_latent_avg:
        if codes.ndim == 2:
            codes = codes + net.latent_avg.repeat(codes.shape[0], 1, 1)[:, 0, :]
        else:
            codes = codes + net.latent_avg.repeat(codes.shape[0], 1, 1)
    if codes.shape[1] == 18 and is_cars:
        codes = codes[:, :16, :]
    return codes



def get_all_latents(net, data_loader, n_images=None, is_cars=False):
    all_latents = []
    i = 0
    with torch.no_grad():
        for batch in data_loader:
            if n_images is not None and i > n_images:
                break
            x = batch
            inputs = x.to(device).float()
            latents = get_latents(net, inputs, is_cars)
            all_latents.append(latents)
            i += len(latents)
    return torch.cat(all_latents)


def save_image(img, save_dir, idx):
    result = tensor2im(img)
    im_save_path = os.path.join(save_dir, f"{idx:05d}.jpg")
    Image.fromarray(np.array(result)).save(im_save_path)


@torch.no_grad()
def generate_inversions(args, g, latent_codes, is_cars):
    print('Saving inversion images')
    inversions_directory_path = os.path.join(args.save_dir, 'inversions')
    os.makedirs(inversions_directory_path, exist_ok=True)
    for i in range(min(args.n_sample, len(latent_codes))):
        imgs, _ = g([latent_codes[i].unsqueeze(0)], input_is_latent=True, randomize_noise=False, return_latents=True)
        if is_cars:
            imgs = imgs[:, :, 64:448, :]
        save_image(imgs[0], inversions_directory_path, i + 1)


def run_alignment(image_path):
    predictor = dlib.shape_predictor(paths_config.model_paths['shape_predictor'])
    aligned_image = align_face(filepath=image_path, predictor=predictor)
    print("Aligned image has shape: {}".format(aligned_image.size))
    return aligned_image


if __name__ == "__main__":
    device = "cuda"

    parser = argparse.ArgumentParser(description="Inference")
    parser.add_argument("--images_dir", type=str, default=None,
                        help="The directory of the images to be inverted")
    parser.add_argument("--save_dir", type=str, default=None,
                        help="The directory to save the latent codes and inversion images. (default: images_dir")
    parser.add_argument("--batch", type=int, default=1, help="batch size for the generator")
    parser.add_argument("--n_sample", type=int, default=None, help="number of the samples to infer.")
    parser.add_argument("--latents_only", action="store_true", help="infer only the latent codes of the directory")
    parser.add_argument("--align", action="store_true", help="align face images before inference")
    parser.add_argument("ckpt", metavar="CHECKPOINT", help="path to generator checkpoint")

    args = parser.parse_args()
    main(args)

**Comment by guligualagulu** [comment]
> @guligualagulu can you guide me how I would do that



> @kwanyun it is [4, 18, 512] does this mean it has latent space representation fo 4 images?

You are right, so I set batchsize to 1 in my code.

**Comment by bikraj2** [comment]
@guligualagulu Thank you so much for your response. I will test out your code now.

**Comment by guligualagulu** [comment]
> @guligualagulu Thank you so much for your response. I will test out your code now.

As you can see, my code is messed up, I will upload the Google Cloud Drive connection later.

**Comment by guligualagulu** [comment]
> @guligualagulu Thank you so much for your response. I will test out your code now.

https://drive.google.com/file/d/1XyVJzNk9lWGf4H3JlJv3kATpMmX7Gn57/view?usp=drive_link
Hope it works.

**Comment by bikraj2** [comment]
@guligualagulu I have made a request for access

**Comment by guligualagulu** [comment]
> @guligualagulu I have made a request for access

I have accepted. Hope everything goes well.

**Comment by guligualagulu** [comment]
> @guligualagulu I have made a request for access

This is the command I tested.
CUDA_VISIBLE_DEVICES=1 python scripts/inference.py \
--images_dir=./datasets/total/test \
--save_dir=./datasets/total/result \
--latents_only \
./total2/directory/checkpoints/best_model.pt

**Comment by bikraj2** [comment]
Thank you so much.The code you provided works perfectly.I just have one more issue on this the inverted images are for some reason different than what I provided.Is it due to the encoder it is using?

**Comment by guligualagulu** [comment]
> Thank you so much.The code you provided works perfectly.I just have one more issue on this the inverted images are for some reason different than what I provided.Is it due to the encoder it is using?

I found this problem too. It seems to me that e4e uses the reconstructed image to obtain the image encoding. This results in a difference between the image encoding and the original image.

**Comment by bikraj2** [comment]
> > Thank you so much.The code you provided works perfectly.I just have one more issue on this the inverted images are for some reason different than what I provided.Is it due to the encoder it is using?
> 
> I found this problem too. It seems to me that e4e uses the reconstructed image to obtain the image encoding. This results in a difference between the image encoding and the original image.

How were you able to fix it?Or did you train an encoder of your own?

**Comment by guligualagulu** [comment]
> > > Thank you so much.The code you provided works perfectly.I just have one more issue on this the inverted images are for some reason different than what I provided.Is it due to the encoder it is using?
> > 
> > 
> > I found this problem too. It seems to me that e4e uses the reconstructed image to obtain the image encoding. This results in a difference between the image encoding and the original image.
> 
> How were you able to fix it?Or did you train an encoder of your own?

To be honest, I didn't make any specific improvements for this. I simply used all the training and testing images for e4e training and increased the number of training epochs to obtain relatively better latent codes. However, despite these changes, the generated latent codes still have some slight differences compared to the images.
